Fix: Handling of Multiple Scopes in Okta .NET SDK v9

aniket-okta opened the pull request:
This pull request resolves the issue for incorrect handling of multiple scopes in Okta .NET SDK version 9. The problem originated from the double encoding of the space character in scopes, which was being replaced with a + character during the string join operation and re-encoded by an underlying library. This caused the illegal_custom_scope error when multiple scopes were used.

Changes Made:
1. Fixed the issue in the OAuthAPI by replacing the + character with a space during scope processing.
2. Added a new test case to validate the handling of multiple scopes. 
3. Updated the README file to reflect accurate usage examples with HashSet.

Issue Analysis:
- Compared logs from v7.0.0 and v9.0.0 to identify the root cause.

bryanapellanes-okta commented on src/Okta.Sdk.IntegrationTest/OAuthScenarios.cs:
```diff
@@ -953,7 +953,99 @@ public async Task ListUsersWithPaginationAndRetry()
+                // Add apps.read scope grant to the service
+                var appsReadGrantResponse = await apiClient.PostAsync<JObject>($"/api/v1/apps/{clientId}/grants", requestOptions);
+
+                // Use OAuth to get list of users
```
I don't think this comment is accurate

bryanapellanes-okta commented on src/Okta.Sdk.IntegrationTest/OAuthScenarios.cs:
```diff
@@ -953,7 +953,99 @@ public async Task ListUsersWithPaginationAndRetry()
+
+                var tokenResponse = await oauthApi.GetBearerTokenAsync();
+
+                tokenResponse.Should().NotBeNull();
```
I think there should be a check for success here.

laura-rodriguez replied:
I agree with Bryan. If we checked grants individually before, let's check again to see if we can retrieve users and apps with this token.

bryanapellanes-okta commented on src/Okta.Sdk.IntegrationTest/OAuthScenarios.cs:
```diff
@@ -953,7 +953,99 @@ public async Task ListUsersWithPaginationAndRetry()
+                requestOptions.Data = JObject.Parse(appsReadGrantPayload);
 
+                // Add apps.read scope grant to the service
+                var appsReadGrantResponse = await apiClient.PostAsync<JObject>($"/api/v1/apps/{clientId}/grants", requestOptions);
```
We should ensure that the grant responses succeed before continuing.

laura-rodriguez replied:
Can we also add a comment on how we're structuring this test? It seems we first test individual grants separately before testing multiple grants. It will be helpful to have some comments if we have to revisit this in the future.

bryanapellanes-okta commented on src/Okta.Sdk/Api/OAuthApi.cs:
```diff
@@ -169,7 +169,7 @@ public async Task<ApiResponse<OAuthTokenResponse>> GetBearerTokenWithHttpInfoAsy
             _dpopProofJwtGenerator.RotateKeys();
             var jwtSecurityToken = _clientAssertionJwtGenerator.GenerateJwt();
-            var scopes = string.Join("+", Configuration.Scopes);
+            var scopes = string.Join(" ", Configuration.Scopes);
```
Since this file is generated, we'll need to update the templates and re-generate the sdk using the openapi-generator.

bryanapellanes-okta reviewed (commented):
A few comments to address.

aniket-okta commented:
@laura-rodriguez @bryanapellanes-okta 
The following changes have been made based on the review comments:

- Enhanced Comments: Updated comments to make it more general and provide a clearer understanding of the test's purpose.
- Token Management: Ensured proper token retrieval and usability for scope operations involving users and apps.
- Grant Responses: Verified grant responses before proceeding to the next step.
- Template Modification: Modified the template to ensure accurate SDK generation.

Please let me know if any other changes are required and if these changes are acceptable.

laura-rodriguez commented on CHANGELOG.md:
```diff
@@ -1,6 +1,10 @@
 
+## 9.0.4
+### Fixed
+- Handling of Multiple Scopes in Okta .NET SDK v9
```
If we have a GH issue for this fix, let's include it here. For example:
- Handling of Multiple Scopes in Okta .NET SDK v9 (#xy)

aniket-okta replied:
@laura-rodriguez it was not there for this particular task.

laura-rodriguez replied:
IIRC the following issue is related to this fix. Take a look at the comments: https://github.com/okta/okta-sdk-dotnet/issues/753

laura-rodriguez commented on src/Okta.Sdk/Okta.Sdk.csproj:
```diff
@@ -13,7 +13,7 @@
     <Copyright>Okta, Inc.</Copyright>
     <RootNamespace>Okta.Sdk</RootNamespace>
-    <Version>9.0.3</Version>
+    <Version>9.0.4</Version>
```
Since this SDK is auto-generated, you will have to update the package version in the [openapi3/config.json file](https://github.com/okta/okta-sdk-dotnet/blob/master/openapi3/config.json#L9). Once you change the version there, go ahead and regenerate the SDK (check [this](https://github.com/okta/okta-sdk-dotnet/blob/master/openapi3/README.md) out). Verify that the project's version and the OauthApi file look OK.

aniket-okta replied:
@laura-rodriguez I've updated the version in config.json and made the corresponding updates to the changelog with the GitHub issue. Additionally, I've regenerated the SDK using the OpenAPI Generator, and everything works as expected. Please review and let me know if everything is working on your end or if there's anything else we need to adjust.

laura-rodriguez reviewed (changes requested):
Thanks for addressing the provided feedback, @aniket-okta. Before merging, we still need to make additional changes related to the codegen process. Please take a look at my comments. Thank you!

bryanapellanes-okta reviewed (approved):
:shipit:

laura-rodriguez reviewed (approved):
LGTM
